lib/std/channel.nim

```diff
@@ -0,0 +1,703 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2021 Mratsim & Nim Contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+
+# Based on https://github.com/mratsim/weave/blob/5696d94e6358711e840f8c0b7c684fcc5cbd4472/unused/channels/channels_legacy.nim
+
+
+##
+## .. warning:: This module is experimental and its interface may change.
+##
+## The following is a simple example of two different ways to use channels:
+## blocking and non-blocking.
+## 
+
+runnableExamples("-r:off --threads:on --gc:orc"):
+  # Be sure to compile with `--threads:on --gc:orc`.
+  import std/os
+
+  # In this example a channel is declared at module scope.
+  # Channels are generic, and they include support for passing objects between
+  # threads.
+  var chan = initChan[string]()
+
+  # This proc will be run in another thread using the threads module.
+  proc firstWorker() =
+    chan.send("Hello World!")
+
+  # This is another proc to run in a background thread. This proc takes a while
+  # to send the message since it sleeps for 2 seconds (or 2000 milliseconds).
+  proc secondWorker() =
+    sleep(2000)
+    chan.send("Another message")
+
+  # Launch the worker.
+  var worker1: Thread[void]
+  createThread(worker1, firstWorker)
+
+  # Block until the message arrives, then print it out.
+  var dest = ""
+  chan.recv(dest)
+  assert dest == "Hello World!"
+
+  # Wait for the thread to exit before moving on to the next example.
+  worker1.joinThread()
+
+  # Launch the other worker.
+  var worker2: Thread[void]
+  createThread(worker2, secondWorker)
+  # This time, use a non-blocking approach with tryRecv.
+  # Since the main thread is not blocked, it could be used to perform other
+  # useful work while it waits for data to arrive on the channel.
+  while true:
+    var msg = ""
+    let tried = chan.tryRecv(msg)
+    if tried:
+      echo msg # "Another message"
```

Comment:
replace all echo's by something we can test, eg a global:
`var witness: int` (with a lock) or an atomic

Reply:
make sense

Reply:
if using a global seq[string] (mimicking echo and collecting lines 1 by 1), you can test with `sort`

Reply:
> The program should output something similar to this, but keep in mind that exact results may vary in the real world::

Reply:
but after `sort` we should have a deterministic output?

Reply:
not really, the length of `messages` may vary.

Reply:
I enable the test with non-strict assert.